### PR TITLE
Tests for SpecOpen and rename restriction to restrict

### PR DESCRIPTION
**Pull request by simonbrandhorst**

### CODE REVIEW

**Comment by simonbrandhorst** [comment]
I marked the failing tests with `@test_broken` to be fixed in separate PRs.

File: experimental/Schemes/SpecOpen.jl [hunk 1]
```diff
@@ -975,7 +1004,7 @@ function maximal_extension(
 end
 
 ### the restriction of a morphism to open subsets in domain and codomain
-function restriction(
```
Comment (simonbrandhorst) [inline experimental/Schemes/SpecOpen.jl:978]: In line 888 we have
`function restrict(f::SpecOpenMor, U::SpecOpen, V::SpecOpen; check::Bool=true)`
so this seems to overwrite the other one.
As far as I can tell each function works sometimes but not always.
Please debug @HechtiDerLachs 
(I tried but failed in reasonable time)

File: experimental/Schemes/SpecOpen.jl [hunk 2]
```diff
@@ -653,7 +653,22 @@ the numerator and denominator of ``f`` have to be elements of
...
 function maximal_extension(
-    X::AbsSpec{<:Ring, <:AbsLocalizedRing}, 
+    X::AbsSpec{<:Ring, <:MPolyLocalizedRing}, 
+    f::AbstractAlgebra.Generic.Frac{RET}
+  ) where {RET<:RingElem}
+
+  a = numerator(f)
+  b = denominator(f)
+  W = OO(X)
+  I = quotient(ideal(W, b), ideal(W, a))
```
Comment (simonbrandhorst) [inline experimental/Schemes/SpecOpen.jl]: Can you explain why this is the way to go? Can't we just do a gcd in the ambient polynomial rings?

File: test/Schemes/SpecOpen.jl [hunk 3]
```diff
@@ -25,9 +25,10 @@ end
   Y = subscheme(X, I)
   Xx = hypersurface_complement(X, x)
   Yx = hypersurface_complement(Y, x)
-  @test_broken complement(Y, Yx)
-  @test_broken is_closed_embedding(Y, Yx)
-  complement(X,Xx)
+  #@test_broken complement(Y, Yx) # Not a valid test! `complement` can only be carried out 
+                                  # whenever the first argument is closed in the second.
```
Comment (simonbrandhorst) [inline test/Schemes/SpecOpen.jl]: Then we should document that. 
Though I think that it makes sense to ask for the complement of an open set. That should return a closed set.
Note also that  we need the second argument closed in the first since we return `complement(X,V) = X \ V = X - V`

File: experimental/Schemes/SpecOpen.jl (hunk 2 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/SpecOpen.jl]: No, unfortunately not. The idea is to produce an ideal that describes the full locus on which `f` is necessarily ill-defined. This follows some parts of Janko's [script](http://www.mathematik.uni-kl.de/~boehm/lehre/17_CA/ca.pdf). In particular, this `quotient` makes a difference when `W` itself is a quotient ring or a localization thereof.

Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: Exactly. But here `X::AbsSpec{<:Ring, <:MPolyLocalizedRing}` and so `W` should be `MPolyLocalizedRing` 
hence there is no quotient and we are working in some localization of a unique factorization domain and there everything should be easy?

File: test/Schemes/SpecOpen.jl (hunk 3 above)
Comment (HechtiDerLachs) [reply test/Schemes/SpecOpen.jl]: Yes, I was thinking so, too. The problem is that `complement` is a rather generic function and one is not advised to document each of its methods. So I am not sure what is the best way to do that. Maybe a more informative error message is the way to go?

As for the complement of an open set: I disagree. The point is that the complement is not well defined as a scheme; it can have arbitrary multiplicities. For the type `PrincipalOpenSubset`, on the other hand, we have a known complement due to its history and in this case I would implement that function.

File: experimental/Schemes/SpecOpen.jl (hunk 2 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/SpecOpen.jl]: Oh. I didn't pay attention to that, sorry. Yes, then what you say probably makes sense. Can you change that? Or should I do it? We might even consider returning a function on a `PrincipalOpenSubset` in this case.

File: test/Schemes/SpecOpen.jl (hunk 3 above)
Comment (simonbrandhorst) [reply test/Schemes/SpecOpen.jl]: Okay convinced. Well if our method has some special behavior in this case we should document it. 
For example there are many docstrings for `sub`. 
I think it is allowed to document the methods separately if the mathematics behind differs. 
Just don't document `Int` vs `fmpz` twice :-)

File: experimental/Schemes/SpecOpen.jl [hunk 4]
```diff
@@ -521,14 +545,16 @@ function generic_fraction(a::SpecOpenRingElem, U::SpecOpen)
   return lifted_numerator(b)//lifted_denominator(b)
 end
 
-(R::MPolyQuoLocalizedRing)(f::SpecOpenRingElem) = restrict(f, Spec(R))
+#(R::MPolyQuoLocalizedRing)(f::SpecOpenRingElem) = restrict(f, Spec(R))
+#(R::MPolyLocalizedRing)(f::SpecOpenRingElem) = restrict(f, Spec(R))
+#(R::MPolyRing)(f::SpecOpenRingElem) = restrict(f, Spec(R))
```
Comment (simonbrandhorst) [inline experimental/Schemes/SpecOpen.jl:546]: Removed these

File: experimental/Schemes/SpecOpen.jl [hunk 5]
```diff
@@ -49,6 +49,10 @@ the list ``f₁,…,fᵣ`` as the *generators* for ``U``.
   end
 end
 
+function set_name!(U::SpecOpen, name::String)
+  U.name = name
+end
+
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl:55]: There is a predefined macro to handle these things. I have to look it up. But @fingolfin suggested to make coherent use of these, so we should probably adapt that.

File: experimental/Schemes/SpecOpen.jl [hunk 6]
```diff
@@ -145,15 +156,28 @@ function SpecOpen(X::AbsSpec, I::MPolyIdeal; check::Bool=true)
...
 end
 
+@doc Markdown.doc"""
+  complement(X::Scheme,Y::Scheme) -> Scheme
+
+Return complement `X \ Y` of `Y` in `X`.
+
+Since we want the complement `U = X \ Y` to be a Scheme,
+we require that `Y` is closed in `X`.
+"""
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
Return the complement ``X \ Y`` of ``Y`` in ``X``.

Since we want the complement `U = X \ Y` to have a well defined scheme structure, 
we require that `Y` is closed in `X`.
"""
```

File: experimental/Schemes/SpecOpen.jl [hunk 7]
```diff
@@ -378,7 +402,7 @@ The ring ``R = 𝒪(X, U)`` belongs to a sheaf of rings ``𝒪(X, -)`` and this
 the scheme ``X`` on which ``𝒪`` is defined.
 """
 scheme(R::SpecOpenRing) = R.scheme
-gens(R::SpecOpenRing) = R.(gens(base_ring(OO(scheme(R)))))
+gens(R::SpecOpenRing) = error("not implemented")
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
gens(R::SpecOpenRing) = R.(gens(ambient_ring(scheme(R))))
```
Why would that go away?

File: experimental/Schemes/SpecOpen.jl [hunk 8]
```diff
@@ -630,9 +656,31 @@ the numerator and denominator of ``f`` have to be elements of
...
+  b = denominator(f)
+  g = gcd(a, b)
+  if !isone(g)
+    a = divexact(a, g)
+    b = divexact(b, g)
+    f = parent(f)(a,b)
+  end
+  W = OO(X)
+  I = ideal(W, b)
+  U = SpecOpen(X, I)
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
  U = SpecOpen(X, [b])
```

File: experimental/Schemes/SpecOpen.jl [hunk 9]
```diff
@@ -877,6 +925,13 @@ function pullback(f::SpecOpenMor)
   return f.pullback::Hecke.Map{typeof(OO(codomain(f))), typeof(OO(domain(f)))}
 end
 
+@doc Markdown.doc"""
+    restrict(f::SchemeMor, U::Scheme, V::Scheme; check::Bool=true)
+
+Return the restriction `g: U -> V` of `f` to `U` and `V`.
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
Return the restriction ``g: U → V`` of ``f`` to ``U`` and ``V``.
```

File: experimental/Schemes/AffineSchemes.jl [hunk 10]
```diff
@@ -576,8 +576,75 @@ function is_closed_embedding(X::AbsSpec, Y::AbsSpec)
...
+                                              <:MPolyPowersOfElement}},
+
+    Y::AbsSpec{<:Ring, <:MPolyQuo}
+  )
+  R = ambient_ring(X)
+  R == ambient_ring(Y) || return false
+  for x in inverted_set(OO(X)) 
+    isunit(OO(Y)(x)) || return false
+  end
+  for g in gens(modulus(OO(Y)))
```
Comment (afkafkafk13) [inline experimental/Schemes/AffineSchemes.jl:639]: Is this 'for'-loop more efficient than iszero(ideal(OO(X),[gens(modulus(OO(Y)))])) or why do it this way?

File: experimental/Schemes/SpecOpen.jl [hunk 11]
```diff
@@ -904,18 +966,32 @@ function restrict(f::SpecOpenMor, W::AbsSpec, Y::AbsSpec; check::Bool=true)
...
+  V = intersect(Y, codomain(f), check=check)
+
+  new_maps_on_patches = [restrict(f[i], U[i], Y, check=check) for i in 1:npatches(U)]
+
+  return SpecOpenMor(U, V, new_maps_on_patches, check=check)
+end
+
 @Markdown.doc """
-    compose(f::T, g::T) where {T<:SpecOpenMor}
+    compose(f::SpecMor, g::SpecMor) -> SpecMor
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]: Why did you change `SpecOpenMor` to `SpecMor`? 

Also, we should probably consider removing the documentation for `compose`, since this is merely one method of a much more general function.

File: experimental/Schemes/SpecOpen.jl (hunk 1 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/SpecOpen.jl:978]: That's weird. By now, they both seem to do the same thing and the code below can go. Probably this had sneaked in unnoticed with a faulty merge? Either way: We should be OK now.

Edit: The debugging has already happened? Or are there still issues?

File: experimental/Schemes/SpecOpen.jl [hunk 12]
```diff
@@ -145,15 +156,28 @@ function SpecOpen(X::AbsSpec, I::MPolyIdeal; check::Bool=true)
...
 end
 
 function complement(X::AbsSpec, Z::AbsSpec{<:Ring, <:MPolyQuo})
+  ambient_ring(X) == ambient_ring(Z) || error("X and Z do not compare")
   return SpecOpen(X, modulus(OO(Z)))
 end
 
 function complement(X::AbsSpec, Z::AbsSpec{<:Ring, <:MPolyLocalizedRing})
+  ambient_ring(X) == ambient_ring(Z) || error("X and Z do not compare")
   return subscheme(X, prod(denominators(inverted_set(OO(Z)))))
```
Comment (afkafkafk13) [inline experimental/Schemes/SpecOpen.jl]: Am I right that this only makes sense for localizations at powers of elements?

**Review by HechtiDerLachs — APPROVED** [review]

File: experimental/Schemes/AffineSchemes.jl (hunk 10 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/AffineSchemes.jl:639]: It probably is. This way, ther is no need to create a new ideal, really.

File: experimental/Schemes/SpecOpen.jl (hunk 12 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/SpecOpen.jl]: Yes. And even worse: This is the type of complement that we decided to discard in the above discussion. So it should go away completely.

File: experimental/Schemes/SpecOpen.jl (hunk 8 above)
Comment (afkafkafk13) [inline experimental/Schemes/SpecOpen.jl]: No, you want SpecOpen(X,[b]), don't you?

File: experimental/Schemes/SpecOpen.jl [hunk 13]
```diff
@@ -145,15 +156,28 @@ function SpecOpen(X::AbsSpec, I::MPolyIdeal; check::Bool=true)
...
 end
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
```

**Comment by afkafkafk13** [comment]
Question about tests: Did I simply miss a test where maximal_extension is tested against a given correct result or is one of that kind still missing?

File: experimental/Schemes/SpecOpen.jl (hunk 11 above)
Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: probably a typo

File: experimental/Schemes/SpecOpen.jl (hunk 7 above)
Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: The name gens is odd. They are not generators of this k-Algebra (it may not even be finitely generated). 
They are merely coordinate functions. So maybe we should choose that name.

**Comment by HechtiDerLachs** [comment]
> Question about tests: Did I simply miss a test where maximal_extension is tested against a given correct result or is one of that kind still missing?

That might very well be still missing. Do you have a non-trivial example?

@simonbrandhorst and I also talked about this, yesterday: The implementations for `maximal_extension` and related methods has not yet undergone a rigorous mathematical check. I have built them to the best of my knowledge, following various outlines in @jankoboehm 's script. But we should eventually check (and proof) that they indeed provide glueings for separated schemes.

File: experimental/Schemes/SpecOpen.jl (hunk 7 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/SpecOpen.jl]: Yes, they are not generators, but rather coordinates. But if we choose this other name all of a sudden while the user is used to writing 'gens' for this, it will also be confusing. Besides: gens is already used for localizations and coordinates is taken by the modules. So I would leave it with my suggestion for now.

Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: Well as a user I can tell you that I expect gens to return generators (in some reasonable sense). 
Besides you never used it. So I would rather leave it undefined for now.

Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: but `coordinates(::AbsSpec)` would make sense?

**Comment by jankoboehm** [comment]
We have an MSc student who is working on modular methods for birational geometry. In that work he bumped in particular into the domain question and is pretty much done proving the algorithms and claims from my lecture notes.

File: test/Schemes/SpecOpen.jl [hunk 14]
```diff
@@ -21,26 +23,66 @@ end
...
 
 @testset "SpecOpen_3" begin
   R, (x,y,u,v) = QQ["x", "y", "u","v"]
   A = Spec(R)
   O = subscheme(A, [x,y,u,v])
   U = complement(A, O)
+  @test is_canonically_isomorphic(A, closure(U))
+  Oscar.SpecOpenRing(U)
+  OU = OO(U)
+  @test_throws ErrorException gens(OO(U))
```
Comment (simonbrandhorst) [inline test/Schemes/SpecOpen.jl]:
```suggestion
  gens(OO(U))
```
